Implementation of Dark Bremsstrahlung Custom Process

revering opened the pull request:
#### PR description:
Implemented physics for dark bremsstrahlung and cross section biasing into SimG4Core/CustomPhysics. Overview of implementation method and motivation available from presentation to SIM group: https://indico.cern.ch/event/1162548/contributions/4882745/attachments/2447856/4194591/Revering_DarkBrem_Sim_22_5_20.pdf

Dark Bremsstrahlung is implemented as an optional process in the CustomPhysics module, and an example _cfi file with brief instructions has been added to SimG4Core/CustomPhysics/DarkBrem_SIM_cfi.py. The new process should have no impact on event generation that does not explicitly call it and include the necessary watcher plugin. A location dependent cross section biasing class is also implemented, and as with the process itself is only instantiated through the DBremWatcher plugin.

#### PR validation:

A small sample of events was generated locally in CMSSW_12_5_X_2022-06-09-1100 to verify that the cross section biasing was taking place in the correct areas of the detector and was producing events at a reasonable rate. Large samples have been generated in CMSSW_10_2_6 to validate the process itself and verify that the cross section calculations are functional for both the total cross section and the expected kinematics. The kinematic calculations do not have any reliance on external cmssw modules, so the behavior should be identical between the two versions.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/30508

- This PR adds an extra 28KB to repository

cmsbuild commented:
A new Pull Request was created by  @revering  for master.

It involves the following packages:

- SimG4Core/CustomPhysics (**simulation**)


@cmsbuild, @civanch, @mdhildreth can you please review it and eventually sign? Thanks.
@makortel, @cvuosalo, @rovere, @bsunanda, @fabiocos, @slomeo this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

civanch commented on SimG4Core/CustomPhysics/src/APrimePhysics.cc:
```diff
@@ -0,0 +1,30 @@
+  G4ParticleDefinition* muonplus = G4MuonPlus::MuonPlusDefinition();
+  G4ProcessManager* pmplus = muonplus->GetProcessManager();
+  G4ProcessManager* pmminus = muonminus->GetProcessManager();
+  pmplus->AddProcess(new G4muDarkBremsstrahlung(mgfile), -1, 1, 1);
```
@revering , I would advice (-1, 1, 1) substitute by (-1, 6, 6)

civanch commented:
@revering , sorry, I was not checking your proposal earlier. Now several comments. First of all technical:
1) all source files should be *.cc
2) all lines with std::cout should be removed or substituted by edm::LogVerbatim
3) edm::LogInfo better to substitute by edm::LogVerbatim
4) if a header file is not used in the header it should be removed or moved to the source
5) #include "G4PhysicalConstants.hh" and #include "G4SystemOfUnits.hh" are allowed only in source files, not in headers.
6) if you use "delete" keyword in a header the method should be public
7) Watchers should be in src not inside plugin 

Now more complicate comments, I am not 100% sure what is the best solution.

1) If you implement some extra product in SimProducer, then you need to take into account that watchers and producers are MT only for Run3 in recent CMSSW releases, for Run-2 it will work only in 1 thread runs. If for your analysis the producer is needed strongly it is possible to implement extra products in this way, but you should take care to submit 1 thread job for Run-2 production.

2) biasing approach is relatively new for Geant4 . In 10.4 and 10.7 the code may be different, I am not sure if it is MT safe in previous version. Are you sure that it is the only solution you may use? I would recommend to use simplest approach - use factor, defined in python configuration file as the particle mass. Multiply cross section by this factor - it is a transparent method.

3) I do not fill numbers, what is the spectra of dark matter bremsstrahlung?

3) May it be that this physics does not need G4VEnergyLossProcess interface by G4VEmProcess interface. The energy loss is used for simulation of ionisation - energy transfer to low-energy electrons. Here we have dark

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/30576

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - SimG4Core/CustomPhysics/src/DarkBremXsecBiasingOperator.cxx:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: c823e3acdc5814895415d017356dfbfa88f81206
  - SimG4Core/CustomPhysics/src/XsecBiasingOperator.cxx:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: c823e3acdc5814895415d017356dfbfa88f81206
  - SimG4Core/CustomPhysics/plugins/DBremWatcher.cc:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: 70e0d61d1f7e6b64a4d2d006d2e38074c22616b1

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/30576/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/30576/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/30577

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - SimG4Core/CustomPhysics/src/DarkBremXsecBiasingOperator.cxx:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: c823e3acdc5814895415d017356dfbfa88f81206
  - SimG4Core/CustomPhysics/src/XsecBiasingOperator.cxx:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: c823e3acdc5814895415d017356dfbfa88f81206
  - SimG4Core/CustomPhysics/plugins/DBremWatcher.cc:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: 70e0d61d1f7e6b64a4d2d006d2e38074c22616b1

cmsbuild commented:
Pull request #38329 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

revering commented:
@civanch Thank you for the comments! I've modified the implementation to hopefully fix the technical issues detailed in notes 1-7, and changed the process to inherit from G4VEmProcess instead of G4VEnergyLossProcess as suggested in note 4. I produced a small sample of 100 events, and the process still occurs around the expected rates.

The main issue remaining is the question of the biasing and multi-threading. While we certainly could implement a global multiplicative factor as you suggest, we would need to re-implement the location dependence and would lose the track weights produced by the biasing operator, limiting the bias factors we could apply without producing non-physical modification to the event locations. I'm not really sure how we could put the location dependence in - my best idea for now would be to modify SampleSecondaries() in the model so that it gets the detector location when an interaction is proposed, and then chooses whether or not to simulate the interaction depending on a given list of allowed detector regions. 
I would really prefer to use the existing bias method if possible - we need the sim-level information about the process that is produced with DBremWatcher for our analysis, so we're already limited to single-threaded running in run 2 generation and we've seen that the biasing works alright there. Is there some kind of test we could run to see if the biasing works in multithreaded generation for run 3?

In addition to the concerns about multithreading and the bias factors, I'm worried about the process disabling that we do to limit the interaction and multithreading. When SampleSecondaries() is called in G4muDarkBremsstrahlungModel, it gets the process manager and sets the process activation to false. At the start of the next event, DBremWatcher then sets it back to true. Do you know if this will work with multithreading? From what I can tell the processManager is thread-local, but I'm not sure if the activation tables they use are a shared variable or not. If they are shared, then one thread producing a brem might turn off the process for the others, or a thread starting an event could turn it on for a thread that needs it disabled.

civanch commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f56f2d/25571/summary.html
**COMMIT**: d117c9ba708d7af82d062d1037ba10c908bc73da
**CMSSW**: CMSSW_12_5_X_2022-06-16-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38329/25571/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3659074
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3659050
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 208 log files, 45 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-06-16-1100+f56f2d/51034/triggerResults): no differences found

civanch commented:
@revering , sorry for the late comment. In part it was because I was not able to comment due to complex method used in implementation of this process.

Just now commit a new PR #38659, in which extra process sub-type is introduced. Please, add in the constructor of your process:

SetProcessSubType(40);

The sub-type is accessible in run time, at any step "creator process" may be checked and this integer code 40 will show that dark bremsstrahlung happens. This for me seems more transparent and clear than weight, which may be changed not only due to biasing but due to other reasons. General biasing method make the scheme too complicate. I personally do not see a reason of that complication, moreover, to me biasing in Geant4 is not well tested. Usually, it is working for specific applications were author take care both on validation and running.

Concerning method overall - the approach you are developing can work in single threading only. For run2 a watcher approach was not MT. For this case, it is OK to run in one thread only.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/31058

- This PR adds an extra 56KB to repository

- Found files with invalid states:
  - SimG4Core/CustomPhysics/src/DarkBremXsecBiasingOperator.cxx:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: c823e3acdc5814895415d017356dfbfa88f81206
  - SimG4Core/CustomPhysics/src/XsecBiasingOperator.cxx:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: c823e3acdc5814895415d017356dfbfa88f81206
  - SimG4Core/CustomPhysics/plugins/DBremWatcher.cc:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: 70e0d61d1f7e6b64a4d2d006d2e38074c22616b1

cmsbuild commented:
Pull request #38329 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

revering commented:
@civanch I changed the process subtype to 40 to match the newly introduced subtype, and I removed the biasing operators and have replaced them with a direct multiplier of the cross section. I generated a small ~50 event sample and it looks like the process occurs at around the expected rate and the process and watcher are correctly seeing the new bias multiplier.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/31061

- This PR adds an extra 72KB to repository

- Found files with invalid states:
  - SimG4Core/CustomPhysics/src/DarkBremXsecBiasingOperator.cc:
    - Added: c823e3acdc5814895415d017356dfbfa88f81206
    - Modified: 2058403331cbfb5e6d6dbdc1df5ce6f748f55d0d, d117c9ba708d7af82d062d1037ba10c908bc73da
    - Deleted: 6cfc15cf8c4d38c0288772315a19727965f04c72
  - SimG4Core/CustomPhysics/interface/XsecBiasingOperator.h:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Modified: 65fd8f1545ad6c69e67987304d8935e4839d27d7
    - Deleted: 6cfc15cf8c4d38c0288772315a19727965f04c72
  - SimG4Core/CustomPhysics/src/DarkBremXsecBiasingOperator.cxx:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: c823e3acdc5814895415d017356dfbfa88f81206
  - SimG4Core/CustomPhysics/src/XsecBiasingOperator.cc:
    - Added: c823e3acdc5814895415d017356dfbfa88f81206
    - Modified: 2058403331cbfb5e6d6dbdc1df5ce6f748f55d0d, 65fd8f1545ad6c69e67987304d8935e4839d27d7, d117c9ba708d7af82d062d1037ba10c908bc73da
    - Deleted: 6cfc15cf8c4d38c0288772315a19727965f04c72
  - SimG4Core/CustomPhysics/interface/DarkBremXsecBiasingOperator.h:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Modified: 65fd8f1545ad6c69e67987304d8935e4839d27d7, d117c9ba708d7af82d062d1037ba10c908bc73da
    - Deleted: 6cfc15cf8c4d38c0288772315a19727965f04c72
  - SimG4Core/CustomPhysics/src/XsecBiasingOperator.cxx:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: c823e3acdc5814895415d017356dfbfa88f81206
  - SimG4Core/CustomPhysics/plugins/DBremWatcher.cc:
    - Added: 904e10221e4dde57eff41d590365a1e369d7f26e
    - Deleted: 70e0d61d1f7e6b64a4d2d006d2e38074c22616b1

civanch commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f56f2d/26286/summary.html
**COMMIT**: 6cfc15cf8c4d38c0288772315a19727965f04c72
**CMSSW**: CMSSW_12_5_X_2022-07-16-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38329/26286/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3654053
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3654023
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 208 log files, 45 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-07-16-1100+f56f2d/51655/triggerResults): no differences found

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
- @civanch @revering do I understand correctly from https://github.com/cms-sw/cmssw/pull/38329#issuecomment-1179269782 that this PR needs https://github.com/cms-sw/cmssw/pull/38659 to be also merged?
- @revering could any of your test script be implemented as a unit test, so that the functionalities implemented here can be kept monitored as running at every new IB?
- @revering squashing all commits in this PR into a single one can get rid of the files with invalid github states as reported in https://github.com/cms-sw/cmssw/pull/38329#issuecomment-1184971871

civanch commented:
@perrotta, because it is PR for exotics we usually perform private tests and not adding a new test for each new analysis. #38659 is orthogonal because it defines process sub-type and is really needed only when simulation performed not for the code itself.

@revering , squash is a simple procedure which merge all commits into one allowing better trace modifications. If you have difficulties to do that, then ask for receipt.

revering commented:
@perrotta @civanch I'm trying to squash the commits together, but I'm having difficulties with it. I pulled in changes from the master branch between a few of my commits, so a complete squash would combine my changes with unrelated ones. I tried to do a rebase to move my commits into one group before squashing them, but git is giving me merge conflicts at every single commit when I do so. I'm not too familiar with squashing - is there some better way to do this?

qliphy commented:
ping @revering

revering commented:
I _think_ I've executed the rebase and squash successfully now. The commit history has all of the dbrem sim commits together as one, although there are also six unrelated commits from other cmssw changes. If these are an issue I think my last resort would be to get a fresh CMSSW version, manually copy the files over, then just add those and force push it.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/31262

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/Eras/python/Modifier_run3_nanoAOD_122_cff.py  modified in PR(s): #38749
  - File  Configuration/StandardSequences/python/Eras.py  modified in PR(s): #38280, #38846, #33532, #38749
  - File  PhysicsTools/NanoAOD/python/boostedTaus_cff.py  modified in PR(s): #38726
  - File  PhysicsTools/NanoAOD/python/nano_eras_cff.py  modified in PR(s): #38749
  - File  PhysicsTools/NanoAOD/python/photons_cff.py  modified in PR(s): #38850

cmsbuild commented:
Pull request #38329 was updated. @perrotta, @ssekmen, @gouskos, @qliphy, @civanch, @rappoccio, @mdhildreth, @cmsbuild, @fgolf, @clacaputo, @sbein, @jpata, @mariadalfonso, @fabiocos, @davidlange6 can you please check and sign again.

perrotta commented:
> I _think_ I've executed the rebase and squash successfully now. The commit history has all of the dbrem sim commits together as one, although there are also six unrelated commits from other cmssw changes. If these are an issue I think my last resort would be to get a fresh CMSSW version, manually copy the files over, then just add those and force push it.

Yes, please do so, if you cannot solve it otherwise (or even a brand new PR...)

Commit:
Implemented optional Dark Bremsstrahlung process into custom physics

Removed unneccesary include of G4GenericBiasingProcess

revering commented:
> > I _think_ I've executed the rebase and squash successfully now. The commit history has all of the dbrem sim commits together as one, although there are also six unrelated commits from other cmssw changes. If these are an issue I think my last resort would be to get a fresh CMSSW version, manually copy the files over, then just add those and force push it.
> 
> Yes, please do so, if you cannot solve it otherwise (or even a brand new PR...)

Done. I copied the contents of SimG4Core/CustomPhysics/ into a new CMSSW_12_5_X_2022-07-25-1100 project, then compiled and ran a test generation to make sure that all of the necessary files transferred successfully. The process works as expected, and there were no changed files beyond those modified to introduce the process.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/31263

- This PR adds an extra 8KB to repository

cmsbuild commented:
Pull request #38329 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38329/31264

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #38329 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f56f2d/26448/summary.html
**COMMIT**: d44f06d670c11554de35f1ab2891dbbf7a2a9c49
**CMSSW**: CMSSW_12_5_X_2022-07-25-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38329/26448/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3667670
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3667646
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 210 log files, 47 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-07-25-1100+f56f2d/51797/triggerResults): no differences found

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

civanch commented:
@revering , let us make a backport to 12_4 (it may be run3 MC production release). This backport is straightforward - the same files as in 12_5 may be committed. 

For 10_6 (Run2 MC production) the backport should be done with care, because modified classes may be different from those of 12_5. It is not possible blindly substitute files but it is needed check classes one by one.

perrotta commented:
+1

forthommel commented on SimG4Core/CustomPhysics/interface/APrimePhysics.h:
```diff
@@ -0,0 +1,40 @@
+#ifndef SIMAPPLICATION_APRIMEPHYSICS_H_
+#define SIMAPPLICATION_APRIMEPHYSICS_H 1_
```
Possible multiple include here (`SIMAPPLICATION_APRIMEPHYSICS_H_` vs. `SIMAPPLICATION_APRIMEPHYSICS_H`)

perrotta replied:
Good catch, thank you @forthommel !
I submitted a fix in https://github.com/cms-sw/cmssw/pull/38862

aandvalenzuela commented:
Hello,

It seems this PR broke `G4VECGEOM` IBs https://github.com/cms-sw/cmssw/issues/38935.
Could you please have a look at it?

Many thanks,
Andrea.

revering commented on SimG4Core/CustomPhysics/interface/G4muDarkBremsstrahlung.h:
```diff
@@ -0,0 +1,40 @@
+
+  G4bool IsApplicable(const G4ParticleDefinition& p) override;
+
+  void PrintInfo() override;
```
I've made a new PR #38949 with PrintInfo() removed to resolve the issue.
